File: jax/21-05/21-05-23/c_star_sky.py
import io,os
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline

def getInts():
    return map(int, input().split())

n, q, c = getInts()

sky = [[[0 for _ in range(101)] for _ in range(101)] for _ in range(11)]

sum = [[[0 for _ in range(101)] for _ in range(101)] for _ in range(11)]

for _ in range(n):
    x, y, s = getInts()
    sky[s][x][y] += 1


def calc(id: int):
    for i in range(1, 101):
        for j in range(1, 101):
            # 带眼睛写
            sum[id][i][j] = sum[id][i - 1][j] + sum[id][i][j - 1] + sky[id][i][j] - sum[id][i - 1][j - 1]

for i in range(c + 1):
    calc(i)

for _ in range(q):
    t, x1, y1, x2, y2 = getInts()
    ans = 0
    for id in range(c + 1):
        val = (id + t) % (c + 1)
        cnt = sum[id][x2][y2] - sum[id][x1 - 1][y2] - sum[id][x2][y1 - 1] + sum[id][x1 - 1][y1 - 1]
        ans += val * cnt
    print(ans)

File: jax/21-09/12/c_star_sky.cc
#include <bits/stdc++.h>
using namespace std;

const int N = 105;

int cnt[N][N][11];
int sum[N][N][11];

int main() {
    int n, q, c;
    cin >> n >> q >> c;

    for (int i = 0; i < n; ++i) {
        int x, y, s;
        cin >> x >> y >> s;
        cnt[x][y][s]++;
    }

    for (int s = 0; s <= c; ++s) {
        for (int i = 1; i <= 100; ++i) {
            for (int j = 1; j <= 100; ++j) {
                sum[i][j][s] = sum[i - 1][j][s] + sum[i][j - 1][s] - sum[i - 1][j - 1][s] + cnt[i][j][s];
            }
        }
    }

    while (q--) {
        int t, x1, y1, x2, y2;
        cin >> t >> x1 >> y1 >> x2 >> y2;
        int ans = 0;

        for (int s = 0; s <= c; ++s) {
            int val = (t + s) % (c + 1);

            int t = sum[x2][y2][s] + sum[x1 - 1][y1 - 1][s] - sum[x1 - 1][y2][s] - sum[x2][y1 - 1][s];

            // cout << "val = " << val << " t = " << t << endl;

            ans += val * t;
        }

        cout << ans << endl;
    }
}